### PR TITLE
move all window hints to the function arguments

### PR DESCRIPTION
**Pull request by sjkelly**
I think this is a better solution than #14 for the time being.

### CODE REVIEW

**Comment by SimonDanisch** [comment]
Maybe we can add some more keywords, e.g. `context_hints`? I'd like to have the context hints decoupled from the rest, since it's really tricky to get them right, and there is usually not much to do differently when configured correctly.
On the other hand, all the window decorations, framebuffer setup etc, are more application dependant.

**Comment by sjkelly** [comment]
So would 
```
                (GLFW.CONTEXT_VERSION_MAJOR, 3),
                (GLFW.CONTEXT_VERSION_MINOR, 3),
```
go in `context_hints`? I'd like to make it possible to dispatch on the GL version and set these as type parameters in screen. I am still getting a feel for OpenGL so I am not completely convinced this is a good idea yet. Nor do I have a design in mind. In the interim, what ever gets GLWindow working easily on versions <3.3 is good to me :)

**Comment by SimonDanisch** [comment]
I mean these:
```
(GLFW.CONTEXT_VERSION_MAJOR, 3),
(GLFW.CONTEXT_VERSION_MINOR, 3),
(GLFW.OPENGL_FORWARD_COMPAT, GL_TRUE),
(GLFW.OPENGL_PROFILE, GLFW.OPENGL_CORE_PROFILE)
```
As you already experienced, these are very tricky to configure to consistently get a similar OpenGL version across platforms.
Which methods would you like to dispatch on the version?
I wanted to have a context object at some point on which to dispatch on, since it's not only the OpenGL version, but also the GLSL version. It gets pretty messy... I mean it already starts messy, since you can't just say "always give me opengl 3.2", and it doesn't get much better from there ;)

**Comment by sjkelly** [comment]
I agree splitting them up makes sense. This should be good to go now!

File: src/reactglfw.jl [hunk 1]
```diff
@@ -377,11 +377,19 @@ function scaling_factor(window::Rectangle{Int}, fb::Vec{2, Int})
...
-function createwindow(name::AbstractString, w, h; debugging = false, windowhints=[(GLFW.SAMPLES, 4)])
+function createwindow(name::AbstractString, w, h; debugging = false,
+                      windowhints=[(GLFW.SAMPLES, 4)],
+                      contexthints=[(GLFW.CONTEXT_VERSION_MAJOR, 3),
+                                    (GLFW.CONTEXT_VERSION_MINOR, 3),
+                                    (GLFW.OPENGL_FORWARD_COMPAT, GL_TRUE),
+                                    (GLFW.OPENGL_PROFILE, GLFW.OPENGL_CORE_PROFILE)])
 
     for elem in windowhints
         GLFW.WindowHint(elem...)
```
Comment (SimonDanisch) [inline src/reactglfw.jl]: maybe `vcat(windowhints, contexthints)` would be better?

Comment (sjkelly) [inline src/reactglfw.jl]: it is definitely nicer syntax, but it would re-allocate the objects.

**Comment by SimonDanisch** [comment]
This doesn't incorperate #14, does it?

**Comment by sjkelly** [comment]
I think it supplants it. I closed it in favor of this one.

**Comment by SimonDanisch** [comment]
Ah okay...because it doesn't really incorperate the solution ;) I guess you just do this in your own package via the context hints then?
That's actually what I would like to avoid. I want to gather all the tricky annoying fixes for cross-plattform variances in one package, which would be GLWindow.
So that you can simply say `createwindow()` instead of:
````julia
@linux_only createwindow(contexthints=xxxx)
@windows_only createwindow(contexthints=yyy)
@osx_only createwindow(contexthints=zzz)
``` 
Which everyone would then need to do in their own package, which comes with bad bug fix propagation.
We could just do something like this:
```Julia
@linux?
const contexthints = [ ... ] 
: @windows?
const contexthints = [ ... ]
: @osx?
const contexthints = [ ... ] 
: error("Unknown Plattform")

createscreen(x; contexthints=contexthints)
```
If that fits your bill ;)

**Comment by sjkelly** [comment]
@SimonDanisch I like that proposal! Can we merge this and add it in another PR? I'll open an issue. I feel like this will take some fiddling to get right on all platforms. In the interim setting `contexthints=[]` works well for me. I promise not to propagate it any further than necessary :P.

**Comment by sjkelly** [comment]
Rebased, and integrated global context hints.

**Comment by SimonDanisch** [comment]
Sorry, I needed to change this PR too much, so I will close this!

If you need core or any profile doesn't depend on the fact if you're on linux or not. With the appropriate drivers on linux you can get OpenGL 4.4 core profile perfectly fine.

>Jessie ship with Mesa <= 10.3 and only support up to OpenGL 3.0 for Intel hardware

I'm not a hundred percent sure, but I think even that is not true. I've had OpenGL 3.3 with a low Mesa version and intel hd 4400 for a pretty long time now (I'm now on Ubuntu 15.10 with Mesa 11.0.2, so it's too late to check)!

I'm addressing this in #25 with https://github.com/JuliaGL/GLWindow.jl/commit/98991d8b262477a419b795f2400c8f12ca2cfcca

**Comment by sjkelly** [comment]
Sounds good to me! I'm planning to update my laptop soon anyway so
hopefully it will have 3.3 then.

**Pull request closed without merging**